File: apps/java/src/test/java/algorithm/LinkedListTest.java
package algorithm;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

class LinkedListTest {

    @Nested
    class SinglyLinkedListTest {
        private SinglyLinkedList<Integer> lst;

        @BeforeEach
        void setUp() { lst = new SinglyLinkedList<>(); }

        @Test void 初期状態は空() { assertEquals(0, lst.size()); assertTrue(lst.isEmpty()); }
        @Test void addFirst() { lst.addFirst(1); assertEquals(1, lst.size()); }
        @Test void addLast() { lst.addLast(1); lst.addLast(2); assertEquals(2, lst.size()); }

        @Test void 探索_見つかる() {
            lst.addLast(10); lst.addLast(20); lst.addLast(30);
            assertTrue(lst.contains(20));
        }
        @Test void 探索_見つからない() {
            lst.addLast(10);
            assertFalse(lst.contains(99));
        }

        @Test void removeFirst() {
            lst.addLast(1); lst.addLast(2);
            lst.removeFirst();
            assertEquals(1, lst.size());
            assertFalse(lst.contains(1));
        }

        @Test void removeLast() {
            lst.addLast(1); lst.addLast(2);
            lst.removeLast();
            assertEquals(1, lst.size());
            assertFalse(lst.contains(2));
        }

        @Test void removeByValue() {
            lst.addLast(1); lst.addLast(2); lst.addLast(3);
            assertTrue(lst.remove(2));
            assertFalse(lst.contains(2));
            assertEquals(2, lst.size());
        }

        @Test void contains() {
            lst.addLast(42);
            assertTrue(lst.contains(42));
            assertFalse(lst.contains(0));
        }

        @Test void removeFirst_空で例外() {
            assertThrows(SinglyLinkedList.EmptyException.class, () -> lst.removeFirst());
        }

        @Test void removeLast_空で例外() {
            assertThrows(SinglyLinkedList.EmptyException.class, () -> lst.removeLast());
        }

        @Test void イテレーション() {
            lst.addLast(1); lst.addLast(2); lst.addLast(3);
            assertEquals(List.of(1, 2, 3), lst.toList());
        }

        @Test void clear() {
            lst.addLast(1); lst.addLast(2);
            lst.clear();
            assertTrue(lst.isEmpty());
        }

        @Test void removeLast_単一要素() {
            lst.addLast(1);
            lst.removeLast();
            assertTrue(lst.isEmpty());
        }

        @Test void remove_空リスト() {
            assertFalse(lst.remove(1));
        }

        @Test void remove_存在しない値() {
            lst.addLast(1); lst.addLast(2);
            assertFalse(lst.remove(99));
            assertEquals(2, lst.size());
        }
    }

    @Nested
    class DoublyLinkedListTest {
        private DoublyLinkedList<Integer> lst;

        @BeforeEach
        void setUp() { lst = new DoublyLinkedList<>(); }

        @Test void 初期状態は空() { assertEquals(0, lst.size()); assertTrue(lst.isEmpty()); }
        @Test void addFirst() { lst.addFirst(1); assertEquals(1, lst.size()); }
        @Test void addLast() { lst.addLast(1); lst.addLast(2); assertEquals(2, lst.size()); }

        @Test void contains() {
            lst.addLast(10); lst.addLast(20);
            assertTrue(lst.contains(20));
            assertFalse(lst.contains(99));
        }

        @Test void remove() {
            lst.addLast(1); lst.addLast(2); lst.addLast(3);
            assertTrue(lst.remove(2));
            assertFalse(lst.contains(2));
            assertEquals(2, lst.size());
        }

        @Test void イテレーション() {
            lst.addLast(1); lst.addLast(2); lst.addLast(3);
            assertEquals(List.of(1, 2, 3), lst.toList());
        }

        @Test void clear() {
            lst.addLast(1);
            lst.clear();
            assertTrue(lst.isEmpty());
        }

        @Test void remove_空リスト() {
            assertFalse(lst.remove(99));
        }
    }

    @Nested
    class ArrayLinkedListTest {
        @Test void 初期化() {
            ArrayLinkedList al = new ArrayLinkedList(100);
            assertEquals(0, al.size());
        }

        @Test void addFirst() {
            ArrayLinkedList al = new ArrayLinkedList(100);
            al.addFirst(1); al.addFirst(2);
            assertEquals(2, al.size());
        }

        @Test void addFirst_順序() {
            ArrayLinkedList al = new ArrayLinkedList(100);
            al.addFirst(1); al.addFirst(2); al.addFirst(3);
            assertEquals(3, al.getHeadData());
        }

        @Test void addLast() {
            ArrayLinkedList al = new ArrayLinkedList(100);
            al.addLast(1); al.addLast(2);
            assertEquals(2, al.size());
        }

        @Test void addLast_順序() {
            ArrayLinkedList al = new ArrayLinkedList(100);
            al.addLast(1); al.addLast(2); al.addLast(3);
            assertEquals(1, al.getHeadData());
        }

        @Test void search_見つかる() {
            ArrayLinkedList al = new ArrayLinkedList(100);
            al.addLast(10); al.addLast(20); al.addLast(30);
            assertTrue(al.search(20) >= 0);
        }

        @Test void search_見つからない() {
            ArrayLinkedList al = new ArrayLinkedList(100);
            al.addLast(10);
            assertEquals(-1, al.search(99));
        }

        @Test void removeFirst() {
            ArrayLinkedList al = new ArrayLinkedList(100);
            al.addLast(1); al.addLast(2); al.addLast(3);
            al.removeFirst();
            assertEquals(2, al.size());
            assertEquals(-1, al.search(1));
        }

        @Test void removeFirst_空() {
            ArrayLinkedList al = new ArrayLinkedList(100);
            al.removeFirst(); // 何もしない
            assertEquals(0, al.size());
        }

        @Test void 削除スロットの再利用() {
            ArrayLinkedList al = new ArrayLinkedList(100);
            al.addFirst(1); al.addFirst(2);
            al.removeFirst();
            al.addFirst(3);
            assertEquals(2, al.size());
        }

        @Test void 容量超過() {
            ArrayLinkedList al = new ArrayLinkedList(2);
            al.addFirst(1); al.addFirst(2);
            al.addFirst(3); // 容量超過で無視
            assertEquals(2, al.size());
        }
    }
}

File: apps/csharp/Algorithm.Tests/LinkedListTest.cs
using Xunit;
using Algorithm;

namespace Algorithm.Tests;

/// <summary>第8章 リスト テスト</summary>
public class LinkedListTest
{
    public class SinglyLinkedListTest
    {
        private readonly SinglyLinkedList<int> _lst = new();

        [Fact] public void 初期状態は空() { Assert.Equal(0, _lst.Size()); Assert.True(_lst.IsEmpty()); }
        [Fact] public void addFirst() { _lst.AddFirst(1); Assert.Equal(1, _lst.Size()); }
        [Fact] public void addLast() { _lst.AddLast(1); _lst.AddLast(2); Assert.Equal(2, _lst.Size()); }
        [Fact] public void 探索_見つかる() { _lst.AddLast(10); _lst.AddLast(20); _lst.AddLast(30); Assert.True(_lst.Contains(20)); }
        [Fact] public void 探索_見つからない() { _lst.AddLast(10); Assert.False(_lst.Contains(99)); }
        [Fact] public void removeFirst() { _lst.AddLast(1); _lst.AddLast(2); _lst.RemoveFirst(); Assert.Equal(1, _lst.Size()); Assert.False(_lst.Contains(1)); }
        [Fact] public void removeLast() { _lst.AddLast(1); _lst.AddLast(2); _lst.RemoveLast(); Assert.Equal(1, _lst.Size()); Assert.False(_lst.Contains(2)); }
        [Fact] public void removeByValue() { _lst.AddLast(1); _lst.AddLast(2); _lst.AddLast(3); Assert.True(_lst.Remove(2)); Assert.False(_lst.Contains(2)); Assert.Equal(2, _lst.Size()); }
        [Fact] public void contains() { _lst.AddLast(42); Assert.True(_lst.Contains(42)); Assert.False(_lst.Contains(0)); }
        [Fact] public void removeFirst_空で例外() => Assert.Throws<SinglyLinkedList<int>.EmptyException>(() => _lst.RemoveFirst());
        [Fact] public void removeLast_空で例外() => Assert.Throws<SinglyLinkedList<int>.EmptyException>(() => _lst.RemoveLast());
        [Fact] public void イテレーション() { _lst.AddLast(1); _lst.AddLast(2); _lst.AddLast(3); Assert.Equal([1, 2, 3], _lst.ToList()); }
        [Fact] public void clear() { _lst.AddLast(1); _lst.AddLast(2); _lst.Clear(); Assert.True(_lst.IsEmpty()); }
        [Fact] public void removeLast_単一要素() { _lst.AddLast(1); _lst.RemoveLast(); Assert.True(_lst.IsEmpty()); }
        [Fact] public void remove_空リスト() => Assert.False(_lst.Remove(1));
        [Fact] public void remove_存在しない値() { _lst.AddLast(1); _lst.AddLast(2); Assert.False(_lst.Remove(99)); Assert.Equal(2, _lst.Size()); }
    }

    public class DoublyLinkedListTest
    {
        private readonly DoublyLinkedList<int> _lst = new();

        [Fact] public void 初期状態は空() { Assert.Equal(0, _lst.Size()); Assert.True(_lst.IsEmpty()); }
        [Fact] public void addFirst() { _lst.AddFirst(1); Assert.Equal(1, _lst.Size()); }
        [Fact] public void addLast() { _lst.AddLast(1); _lst.AddLast(2); Assert.Equal(2, _lst.Size()); }
        [Fact] public void contains() { _lst.AddLast(10); _lst.AddLast(20); Assert.True(_lst.Contains(20)); Assert.False(_lst.Contains(99)); }
        [Fact] public void remove() { _lst.AddLast(1); _lst.AddLast(2); _lst.AddLast(3); Assert.True(_lst.Remove(2)); Assert.False(_lst.Contains(2)); Assert.Equal(2, _lst.Size()); }
        [Fact] public void イテレーション() { _lst.AddLast(1); _lst.AddLast(2); _lst.AddLast(3); Assert.Equal([1, 2, 3], _lst.ToList()); }
        [Fact] public void clear() { _lst.AddLast(1); _lst.Clear(); Assert.True(_lst.IsEmpty()); }
        [Fact] public void remove_空リスト() => Assert.False(_lst.Remove(99));
    }

    public class ArrayLinkedListTest
    {
        [Fact] public void 初期化() => Assert.Equal(0, new ArrayLinkedList(100).Size());
        [Fact] public void addFirst() { var al = new ArrayLinkedList(100); al.AddFirst(1); al.AddFirst(2); Assert.Equal(2, al.Size()); }
        [Fact] public void addFirst_順序() { var al = new ArrayLinkedList(100); al.AddFirst(1); al.AddFirst(2); al.AddFirst(3); Assert.Equal(3, al.GetHeadData()); }
        [Fact] public void addLast() { var al = new ArrayLinkedList(100); al.AddLast(1); al.AddLast(2); Assert.Equal(2, al.Size()); }
        [Fact] public void addLast_順序() { var al = new ArrayLinkedList(100); al.AddLast(1); al.AddLast(2); al.AddLast(3); Assert.Equal(1, al.GetHeadData()); }
        [Fact] public void search_見つかる() { var al = new ArrayLinkedList(100); al.AddLast(10); al.AddLast(20); al.AddLast(30); Assert.True(al.Search(20) >= 0); }
        [Fact] public void search_見つからない() { var al = new ArrayLinkedList(100); al.AddLast(10); Assert.Equal(-1, al.Search(99)); }
        [Fact] public void removeFirst() { var al = new ArrayLinkedList(100); al.AddLast(1); al.AddLast(2); al.AddLast(3); al.RemoveFirst(); Assert.Equal(2, al.Size()); Assert.Equal(-1, al.Search(1)); }
        [Fact] public void removeFirst_空() { var al = new ArrayLinkedList(100); al.RemoveFirst(); Assert.Equal(0, al.Size()); }
        [Fact] public void 削除スロットの再利用() { var al = new ArrayLinkedList(100); al.AddFirst(1); al.AddFirst(2); al.RemoveFirst(); al.AddFirst(3); Assert.Equal(2, al.Size()); }
        [Fact] public void 容量超過() { var al = new ArrayLinkedList(2); al.AddFirst(1); al.AddFirst(2); al.AddFirst(3); Assert.Equal(2, al.Size()); }
    }
}
